src/modules/consumer.js

```diff
@@ -164,50 +199,91 @@ class Consumer {
       // receive message, parse it, execute callback, check if should answer, ack/reject message
       const body = parsers.in(msg);
       try {
+        const shouldContinue = await this.hooks.trigger(this, ConsumerHooks.beforeProcessMessageEvent, {
+          queue,
+          message: msg,
+          content: body,
+        });
+        if (!shouldContinue) {
+          await this._rejectMessageAfterProcess(channel, queue, msg, body, false);
+          return;
+        }
         const res = await callback(body, msg.properties);
         await this.checkRpc(msg.properties, queue, res);
       } catch (error) {
         // if something bad happened in the callback, reject the message so we can requeue it (or not)
-        this._connection.config.logger.error({
+        logger.error({
           message: `${loggerAlias} Failed processing message from queue ${queue}: ${error.message}`,
           error,
           params: { queue, message: messageString },
         });
 
-        try {
-          channel.reject(msg, this._connection.config.requeue);
-        } catch (rejectError) {
-          this._connection.config.logger.error({
-            message: `${loggerAlias} Failed to reject message after processing failure on queue ${queue}: ${rejectError.message}`,
-            error: rejectError,
-            params: { queue },
-          });
-        }
-
+        await this._rejectMessageAfterProcess(channel, queue, msg, body, this._connection.config.requeue, error);
         return;
       }
 
       try {
         channel.ack(msg);
       } catch (ackError) {
-        this._connection.config.logger.error({
+        await this.hooks.trigger(this, ConsumerHooks.afterProcessMessageEvent, {
+          queue,
+          message: msg,
+          content: body,
+          ackError,
+        });
+        logger.error({
           message: `${loggerAlias} Failed to ack message after processing finished on queue ${queue}: ${ackError.message}`,
           error: ackError,
           params: { queue },
         });
+        return;
       }
+      await this.hooks.trigger(this, ConsumerHooks.afterProcessMessageEvent, { queue, message: msg, content: body });
     };
 
     try {
       await channel.consume(queue, consumeFunc, { noAck: false });
     } catch (error) {
-      this._connection.config.logger.error({
+      logger.error({
         message: `${loggerAlias} Failed to start consuming from queue ${queue}: ${error.message}`,
         error,
         params: { queue },
       });
     }
   }
+
+  /** @private */
+  async _rejectMessageAfterProcess(channel, queue, msg, parsedBody, requeue, error) {
+    try {
+      channel.reject(msg, requeue);
+
+      if (!requeue) {
+        // If not requeued and message will be removed from the queue, return rpc error response if needed.
+        await this.checkRpc(msg.properties, queue, error instanceof Error ? { error } : undefined);
```

Comment:
Note: This also fixes an unrelated bug, where if the config has `requeue: false`, the message would be rejected without sending a reply, causing the client to wait until timeout.